backend/lambdas/api/authorizer/authorizer/handlers.py

```diff
@@ -165,25 +172,85 @@ def _verify_claims(claims: dict):
         raise Exception("Unauthorized")
 
 
-def _get_user(email: str) -> User:
-    # Create the user, if necessary
-    user, created = User.objects.get_or_create(email=email, defaults={"scope": DEFAULT_SCOPE})
-    if created:
-        # User was created so create their self group as well
-        Group.create_self_group(user)
-
-    if user.deleted:
-        # User record is soft deleted, return None so that auth fails
-        return None
+@transaction.atomic
+def _get_update_or_create_user(email: str) -> User:
+    """
+    Attempt to get a user based on email.
+    If user does not exist, try to match based on EMAIL_DOMAIN_ALIASES, and update email if successful.
+    If a user is still not found, create a new one.
+    """
+    user = _get_user(email)
 
-    # Set the last login timestamp
+    # If user is found, return the user (unless the user is soft-deleted)
+    if user:
+        # If user is soft-deleted, return None so that auth fails
+        if user.deleted:
+            return None
+        else:
+            return user
+
+    if EMAIL_DOMAIN_ALIASES:
+        email_local_part = email.split("@")[0]
+        email_domain = email.split("@")[1]
+
+        # Check if any aliases exist for domain, and attempt to find a match
+        for alias in EMAIL_DOMAIN_ALIASES:
+            if alias["new_domain"] == email_domain:
+                for old_domain in alias["old_domains"]:
+                    if alias.get("email_transformation"):
+                        transformed_email_local_part = re.sub(
+                            alias["email_transformation"]["new_email_regex"],
+                            alias["email_transformation"]["old_email_expr"],
+                            email_local_part,
+                        )
+                        old_email = f"{transformed_email_local_part}@{old_domain}"
+                    else:
+                        old_email = f"{email_local_part}@{old_domain}"
+
+                    user = _get_user(old_email)
+
+                    # If a user is found with an old email (and was not soft-deleted), update the email and return user
+                    if user and not user.deleted:
+                        user.email = email
+                        user.save()
+                        return user
+
+    # Create the user since no match has been found at this point
+    return _create_user(email)
+
+
+@transaction.atomic
+def _update_login_timestamp(user: User) -> None:
+    """
+    Set the last login timestamp
+    """
     user.last_login = datetime.utcnow().replace(tzinfo=timezone.utc)
     user.save()
 
-    # Return the user
+
+@transaction.atomic
```

Comment:
I thought I would just put the decorator there in case in the future this method was used outside of the other method. I suppose if such a time comes, we can always just add it back.